{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e8bf9d2-279b-4854-9187-15bf19c3c1ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sfaira\n",
    "import os\n",
    "import re\n",
    "import scanpy as sc\n",
    "import warnings\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "warnings.simplefilter('ignore', FutureWarning)\n",
    "warnings.simplefilter('ignore', UserWarning)\n",
    "warnings.simplefilter('ignore', RuntimeWarning)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "91b4eeb2-6247-4c1b-a236-e44dd7dccbed",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"/Users/hassansaei/Desktop/DIR_SFAIRA/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aab538a5-94b9-4283-b977-ca79f03c0b7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total protein_coding & lncRNA genes: 24100\n"
     ]
    }
   ],
   "source": [
    "# Load GTF file\n",
    "gtf_file = \"Homo_sapiens.GRCh38.104.gtf\"\n",
    "genes = []\n",
    "\n",
    "with open(gtf_file, \"r\") as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"#\"):\n",
    "            continue  # Skip headers\n",
    "        columns = line.strip().split(\"\\t\")\n",
    "        if columns[2] == \"gene\":\n",
    "            info = {kv.split(\" \")[0]: kv.split(\" \")[1].strip('\";') for kv in columns[8].split(\"; \") if \" \" in kv}\n",
    "            if info.get(\"gene_biotype\") in [\"protein_coding\", \"lncRNA\"]:\n",
    "                genes.append(info.get(\"gene_name\"))\n",
    "\n",
    "# Create a set of allowed genes\n",
    "allowed_genes = set(genes)\n",
    "allowed_genes = {gene for gene in genes if gene and pd.notna(gene)}\n",
    "print(f\"Total protein_coding & lncRNA genes: {len(allowed_genes)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "07f82b56-4545-4f9d-b0b6-7f88c0e80444",
   "metadata": {},
   "outputs": [],
   "source": [
    "def harmonize_matrix(df):\n",
    "    \"\"\"\n",
    "    Matches gene names in df with the reference gene list (allowed_genes),\n",
    "    keeps common genes, and adds missing ones as zero rows, preserving sparse format.\n",
    "    \"\"\"\n",
    "    # Get existing genes in the dataset\n",
    "    dataset_genes = set(df.index)\n",
    "    # Find genes that exist in both datasets\n",
    "    common_genes = sorted(allowed_genes.intersection(dataset_genes))\n",
    "    # Find missing genes that need to be added with zero values\n",
    "    missing_genes = sorted(allowed_genes - dataset_genes)\n",
    "    # Keep only the common genes\n",
    "    df = df.loc[common_genes, :]\n",
    "    # Create a zero-filled sparse matrix for missing genes\n",
    "    zero_matrix = sp.csr_matrix((len(missing_genes), df.shape[1]))  # Sparse zeros\n",
    "    # Convert missing genes into a sparse DataFrame\n",
    "    zero_df = pd.DataFrame.sparse.from_spmatrix(zero_matrix, index=missing_genes, columns=df.columns)\n",
    "    # Append missing genes with zero values\n",
    "    harmonized_df = pd.concat([df, zero_df])\n",
    "    # Ensure correct gene order (sorted allowed_genes)\n",
    "    harmonized_df = harmonized_df.loc[sorted(allowed_genes), :]\n",
    "    # Transpose and return\n",
    "    return harmonized_df.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "26fa404c-8271-4067-92f5-edc847433394",
   "metadata": {},
   "outputs": [],
   "source": [
    "def memory_efficient_harmonize(input_file, output_file, harmonize_matrix):\n",
    "    \"\"\"\n",
    "    Loads an AnnData object, creates a transposed (and possibly harmonized)\n",
    "    DataFrame in a memory efficient way, then writes a new AnnData file.\n",
    "\n",
    "    Parameters:\n",
    "      input_file (str): Path to the input h5ad file.\n",
    "      output_file (str): Path where the output h5ad will be saved.\n",
    "      harmonize_matrix (callable): A function that takes a DataFrame and returns a harmonized DataFrame.\n",
    "    \"\"\"\n",
    "    # Read the h5ad file\n",
    "    adata = sc.read_h5ad(input_file)\n",
    "    sc.pp.filter_cells(adata, min_genes=300)\n",
    "    sc.pp.filter_genes(adata, min_cells=3)\n",
    "    # Check if the underlying matrix is sparse\n",
    "    if sp.issparse(adata.X):\n",
    "        # Create a sparse DataFrame from the sparse matrix.\n",
    "        # NOTE: We first create the DataFrame in the original orientation,\n",
    "        # then transpose it to avoid doing a dense conversion.\n",
    "        df = pd.DataFrame.sparse.from_spmatrix(\n",
    "            adata.X,\n",
    "            index=adata.obs_names,\n",
    "            columns=adata.var_names\n",
    "        )\n",
    "    else:\n",
    "        # Otherwise, create a regular DataFrame\n",
    "        df = pd.DataFrame(\n",
    "            adata.X,\n",
    "            index=adata.obs_names,\n",
    "            columns=adata.var_names\n",
    "        )\n",
    "    df = harmonize_matrix(df.T)\n",
    "    adata_new = sc.AnnData(\n",
    "        X=df, \n",
    "        obs=adata.obs.copy(), \n",
    "        var=pd.DataFrame(index=df.columns)\n",
    "    )\n",
    "    adata_new.write(output_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b687f48-a2a5-49ef-9a2f-23f9538cc7b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1\n",
    "\n",
    "files = os.listdir(\"raw/dno_doi_kidney_organoid_GSE147393/GSE147393_RAW/\")\n",
    "\n",
    "files = list(set([re.sub(r\"(_barcodes\\.tsv|_matrix\\.mtx|_features\\.tsv)\\.gz\", \"\", x) + \"_\" for x in files]))\n",
    "\n",
    "basepath = \"raw/dno_doi_kidney_organoid_GSE147393/GSE147393_RAW/\"\n",
    "study = \"dno_doi_kidney_organoid_2024_GSE147393\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix = prefix # Ensure prefix is used correctly\n",
    "    )\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [re.search(r'_(.*?)_', x).group(1) if re.search(r'_(.*?)_', x) else x for x in adata.obs.sample_id]\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"Pd_iPSC\" # Patient derived iPSC\n",
    "    adata.obs['diff_protocol'] = \"Howden_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['study_id'] = \"1\"\n",
    "    adata.obs_names_make_unique()\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"null\": \"HNF4A_del\", \"corrected\": \"HNF4A_corrected\", \"patient\": \"HNF4A_Het\"}\n",
    "mapping2 = {\"null\": \"OMIM_600281\", \"corrected\": \"Control\", \"patient\": \"OMIM_600281\"}\n",
    "\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"disease\"] = adata.obs[\"sample_id\"].map(mapping2)\n",
    "adata.obs[\"condition\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/dno_doi_kidney_organoid_GSE147393/dno_doi_kidney_organoid_2024_GSE147393.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87651db7-8e98-4df0-b9eb-bddd325e5f7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file='raw/dno_doi_kidney_organoid_GSE147393/dno_doi_kidney_organoid_GSE147393.h5ad',\n",
    "                           output_file='raw/dno_doi_kidney_organoid_GSE147393/dno_doi_kidney_organoid_GSE147393.h5ad',\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cf5bef1-b04d-47f7-a26f-bb960645e9a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2\n",
    "\n",
    "files = os.listdir(\"raw/d10_1016_j_celrep_2024_114310/GSE230848_RAW/\")\n",
    "\n",
    "files = list(set([re.sub(r\"(.barcodes\\.tsv|.matrix\\.mtx|.features\\.tsv)\\.gz\", \"\", x) + \".\" for x in files]))\n",
    "\n",
    "basepath = \"raw/d10_1016_j_celrep_2024_114310/GSE230848_RAW/\"\n",
    "study = \"d10_1016_j_celrep_2024_114310\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix = prefix # Ensure prefix is used correctly\n",
    "    )\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"HESC\" # Patient derived iPSC\n",
    "    adata.obs['diff_protocol'] = \"Freedman_2015\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d24\"\n",
    "    adata.obs['study_id'] = \"2\"\n",
    "    adata.obs_names_make_unique()\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_celrep_2024_114310/d10_1016_j_celrep_2024_114310.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "38845b3b-143b-44bc-81d1-10891680ab23",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file='raw/d10_1016_j_celrep_2024_114310/d10_1016_j_celrep_2024_114310.h5ad',\n",
    "                          output_file='raw/d10_1016_j_celrep_2024_114310/d10_1016_j_celrep_2024_114310.h5ad',\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "38cead85-9cc0-46bb-a7c2-1d7664c1b170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 3\n",
    "\n",
    "files = os.listdir(\"raw/d10_1016_j_stem_2023_12_003/GSE244608_RAW/\")\n",
    "\n",
    "names = [x.split(\"_\")[0] for x in files] \n",
    "\n",
    "basepath = \"raw/d10_1016_j_stem_2023_12_003/GSE244608_RAW/\"\n",
    "study = \"d10_1016_j_stem_2023_12_003\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_h5(\n",
    "        basepath + prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"Pd_iPSC\" # Patient derived iPSC\n",
    "    adata.obs['diff_protocol'] = \"Low_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_ATAC_Gene\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['study_id'] = \"3\"\n",
    "    adata.obs['type'] = \"snRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"ARPKD\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM7821874\": \"PKHD1\", \"GSM7821876\": \"PKHD1_corrected\", \"GSM7821878\": \"PKHD1 mutation, KIF3A mutation\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"condition\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2023_12_003/d10_1016_j_stem_2023_12_003.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dad3a0cd-a702-4fb9-8a2e-7279f0ce14ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file='raw/d10_1016_j_stem_2023_12_003/d10_1016_j_stem_2023_12_003.h5ad',\n",
    "                          output_file='raw/d10_1016_j_stem_2023_12_003/d10_1016_j_stem_2023_12_003.h5ad',\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6911a262-ee41-4784-a35a-8d132096429a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 4\n",
    "\n",
    "file_path = \"raw/d10_1016_j_stem_2019_06_009/GSE132023_Day10_12_14_count.txt.gz\"\n",
    "\n",
    "# Read count matrix (assuming tab-separated values)\n",
    "df = pd.read_csv(file_path, sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "\n",
    "meta = df.iloc[0]\n",
    "df = df.iloc[1:] \n",
    "\n",
    "adata = sc.AnnData(df.T)\n",
    " \n",
    "adata.obs[\"sample_id\"] = meta.values\n",
    "\n",
    "adata.obs_names_make_unique()\n",
    "adata.var_names_make_unique()\n",
    "adata.obs[\"study\"] = \"d10_1016_j_stem_2019_06_009\"\n",
    "adata.obs['source'] = \"hPSCs\" # Patient derived iPSC\n",
    "adata.obs['diff_protocol'] = \"Low_2019\"\n",
    "adata.obs['sc_protocol'] = \"10x_3_v1\"\n",
    "adata.obs['sequencing'] = \"Illumina_HiSeq_4000 \"\n",
    "adata.obs['genome_build'] = \"hg38\"\n",
    "adata.obs['study_id'] = \"4\"\n",
    "adata.obs['type'] = \"scRNAseq\"\n",
    "adata.obs[\"disease\"] = \"Control\"\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping1 = {\"Day10rep1\": \"GSM3834523\", \"Day10rep2\": \"GSM3834524\", \"Day12\": \"GSM3834525\", \"Day14rep1\": \"GSM3834526\", \"Day14rep2\": \"GSM3834527\"}\n",
    "mapping2 = {\"Day10rep1\": \"d10\", \"Day10rep2\": \"d10\", \"Day12\": \"d12\", \"Day14rep1\": \"d14\", \"Day14rep2\": \"d14\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"sample\"] = adata.obs[\"sample_id\"].map(mapping1)\n",
    "adata.obs[\"day\"] = adata.obs[\"sample_id\"].map(mapping2)\n",
    "\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "# Convert `adata.X` to a proper format\n",
    "if isinstance(adata.X, np.ndarray):\n",
    "    adata.X = np.array(adata.X, dtype=np.float32)  # Ensure float type\n",
    "elif isinstance(adata.X, sp.spmatrix):\n",
    "    adata.X = adata.X.astype(np.float32)  # Ensure correct sparse format\n",
    "else:\n",
    "    raise TypeError(f\"Unexpected type for adata.X: {type(adata.X)}\")\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2019_06_009/d10_1016_j_stem_2019_06_009.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bea62c8c-9246-4854-91e0-7a71d8ff9109",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2019_06_009/d10_1016_j_stem_2019_06_009.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2019_06_009/d10_1016_j_stem_2019_06_009.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65b9d527-23d1-433b-8c20-92e9f2b0cd69",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct feature files\n",
    "import glob\n",
    "import gzip\n",
    "\n",
    "basepath = \"raw/dno_doi_kidney_organoid_GSE241972/GSE241972_RAW/\"\n",
    "feature_files = glob.glob(os.path.join(basepath, \"*_features.tsv.gz\"))\n",
    "\n",
    "# Process each file\n",
    "for file in feature_files:\n",
    "    # Read features file\n",
    "    df = pd.read_csv(file, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Add new column \"Gene Expression\"\n",
    "    df[\"Gene Expression\"] = \"Gene Expression\"\n",
    "\n",
    "    # Write back to gzipped file without adding extra newlines\n",
    "    with gzip.open(file, \"wt\", encoding=\"utf-8\") as f:\n",
    "        df.to_csv(f, sep=\"\\t\", header=False, index=False, lineterminator=\"\\n\")  # Ensures no extra newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8af42eea-4e44-45e0-99d8-4fec106818eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 5\n",
    "\n",
    "files = os.listdir(\"raw/dno_doi_kidney_organoid_GSE241972/GSE241972_RAW/\")\n",
    "\n",
    "files = list(set([re.sub(r\"(_barcodes\\.tsv|_matrix\\.mtx|_features\\.tsv|_genes\\.tsv)\\.gz\", \"\", x) + \"_\" for x in files]))\n",
    "\n",
    "basepath = \"raw/dno_doi_kidney_organoid_GSE241972/GSE241972_RAW/\"\n",
    "study = \"dno_doi_kidney_organoid_GSE241972\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        prefix = prefix # Ensure prefix is used correctly\n",
    "    )\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    #adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"KPC\" # Kidney primary cells\n",
    "    adata.obs['diff_protocol'] = \"Guo_2023\"\n",
    "    adata.obs['sc_protocol'] = \"SMARTER\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d24\"\n",
    "    adata.obs['study_id'] = \"5\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/dno_doi_kidney_organoid_GSE241972/dno_doi_kidney_organoid_GSE241972.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7258067a-50b8-4b10-ac07-2d70cce2819b",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/dno_doi_kidney_organoid_GSE241972/dno_doi_kidney_organoid_GSE241972.h5ad\",\n",
    "                          output_file=\"raw/dno_doi_kidney_organoid_GSE241972/dno_doi_kidney_organoid_GSE241972.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "cb467e36-5207-4791-b4c6-d7102618df45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 6\n",
    "\n",
    "files = os.listdir(\"raw/d10_1038_s41467_022_33623_z/GSE184928_RAW/\")\n",
    "\n",
    "files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1038_s41467_022_33623_z/GSE184928_RAW/\"\n",
    "study = \"d10_1038_s41467_022_33623_z\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        prefix= prefix,\n",
    "        var_names=\"gene_symbols\",\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Vanslambrouck_2022\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d25\"\n",
    "    #adata.obs['study_id'] = \"6\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM5600482\": \"d13\", \"GSM5600483\": \"d27\", \"GSM7821878\": \"d25\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/d10_1038_s41467_022_33623_z/d10_1038_s41467_022_33623_z.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "686177c4-80eb-4364-b74e-9a040853c05c",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1038_s41467_022_33623_z/d10_1038_s41467_022_33623_z.h5ad\",\n",
    "                          output_file=\"raw/d10_1038_s41467_022_33623_z/d10_1038_s41467_022_33623_z.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7b18aaa3-8676-46e3-8cda-7475165edede",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 7\n",
    "\n",
    "files = os.listdir(\"raw/d10_1016_j_stem_2022_06_005/GSE164564_RAW/\")\n",
    "\n",
    "#files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1016_j_stem_2022_06_005/GSE164564_RAW/\"\n",
    "study = \"d10_1016_j_stem_2022_06_005\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath + prefix,\n",
    "        var_names=\"gene_symbols\",\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Ransick_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NextSeq_500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d25\"\n",
    "    #adata.obs['study_id'] = \"6\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM5014155\": \"d8\", \"GSM5014156\": \"d10\", \"GSM5014157\": \"d14\", \"GSM5014158\": \"d16\", \"GSM5014159\": \"d28\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2022_06_005/d10_1016_j_stem_2022_06_005.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e9bfd73-16a1-4ef5-b77d-cd6297c2ded6",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2022_06_005/d10_1016_j_stem_2022_06_005.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2022_06_005/d10_1016_j_stem_2022_06_005.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "8c4ba57e-cfb0-4038-b5f9-2d6d751e37df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 8\n",
    "\n",
    "files = os.listdir(\"raw/d10_1242_dev_200198/GSE181954_RAW/\")\n",
    "\n",
    "files = set([x.split(\"_\")[0] for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1242_dev_200198/GSE181954_RAW/\"\n",
    "study = \"d10_1242_dev_200198\"\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    #adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Jansen_2022\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d24\"\n",
    "    #adata.obs['study_id'] = \"6\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adata.write(\"raw/d10_1242_dev_200198/d10_1242_dev_200198.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ef23cef-aa00-4026-8123-cd176d1a93f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1242_dev_200198/d10_1242_dev_200198.h5ad\",\n",
    "                          output_file=\"raw/d10_1242_dev_200198/d10_1242_dev_200198.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6528b38a-ea60-4b8b-9174-1cf7ee11c62a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from cloupe import Cloupe\n",
    "\n",
    "# parse file\n",
    "c = Cloupe(\"raw/dno_doi_kidney_organoid_GSE164647/GSE164647_Injured_kidney_organoids_gene_expression_profile_at_single_cell.cloupe\", load_csr=True)\n",
    "\n",
    "# extract barcodes, features and projections to CSVs\n",
    "c.to_csv(outdir=\"raw/dno_doi_kidney_organoid_GSE164647\")\n",
    "\n",
    "# create anndata with the same name as input file but `.h5ad` extension\n",
    "c.to_anndata(\"raw/dno_doi_kidney_organoid_GSE164647/dno_doi_kidney_organoid_GSE164647.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "09dc042f-8acb-442e-a0bc-e96d2441d70a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 9\n",
    "adata = sc.read_h5ad(\"raw/dno_doi_kidney_organoid_GSE164647/dno_doi_kidney_organoid_GSE164647.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4217640d-111f-49e5-bb06-8e2e42e1e412",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/dno_doi_kidney_organoid_GSE164647/dno_doi_kidney_organoid_GSE164647.h5ad\",\n",
    "                          output_file=\"raw/dno_doi_kidney_organoid_GSE164647/dno_doi_kidney_organoid_GSE164647.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "17b29fb6-2cd4-49fa-b137-75d840b85a96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 10\n",
    "\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "\n",
    "file_path = \"raw/d10_1093_stmcls_sxac009/GSE165266_RAW/GSM5029240_Sample1.csv\"\n",
    "\n",
    "# Read count matrix (assuming tab-separated values)\n",
    "df = pd.read_csv(file_path, sep=\",\", index_col=0)\n",
    "\n",
    "adata = sc.AnnData(df.T)\n",
    " \n",
    "adata.obs[\"sample_id\"] = \"GSM5029240\"\n",
    "\n",
    "adata.obs_names_make_unique()\n",
    "adata.var_names_make_unique()\n",
    "adata.obs[\"study\"] = \"d10_1093_stmcls_sxac009\"\n",
    "adata.obs['source'] = \"hiPSC\" # iPSC\n",
    "adata.obs['diff_protocol'] = \"Takasato_2016\"\n",
    "adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "adata.obs['genome_build'] = \"hg38\"\n",
    "adata.obs['type'] = \"scRNAseq\"\n",
    "adata.obs[\"disease\"] = \"Control\"\n",
    "\n",
    "# Convert `adata.X` to a proper format\n",
    "if isinstance(adata.X, np.ndarray):\n",
    "    adata.X = np.array(adata.X, dtype=np.float32)  # Ensure float type\n",
    "elif isinstance(adata.X, sp.spmatrix):\n",
    "    adata.X = adata.X.astype(np.float32)  # Ensure correct sparse format\n",
    "else:\n",
    "    raise TypeError(f\"Unexpected type for adata.X: {type(adata.X)}\")\n",
    "\n",
    "adata.write(\"raw/d10_1093_stmcls_sxac009/d10_1093_stmcls_sxac009.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c04b7fe-ad8d-4e85-9cc8-432458dec572",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1093_stmcls_sxac009/d10_1093_stmcls_sxac009.h5ad\",\n",
    "                          output_file=\"raw/d10_1093_stmcls_sxac009/d10_1093_stmcls_sxac009.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "d93b8f1f-7589-4ee8-ace4-7d944acdac7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 11\n",
    "\n",
    "files = os.listdir(\"raw/10_1016_j_stem_2020_12_001/GSE161255_RAW/\")\n",
    "\n",
    "files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/10_1016_j_stem_2020_12_001/GSE161255_RAW/\"\n",
    "study = \"10_1016_j_stem_2020_12_001\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Howden_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d25+14\"\n",
    "    #adata.obs['study_id'] = \"6\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Conmemory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2020_12_001/d10_1016_j_stem_2020_12_001.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2020_12_001/d10_1016_j_stem_2020_12_001.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)trol\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2020_12_001/d10_1016_j_stem_2020_12_001.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "baf74e31-2ba6-4bf3-9cc5-6ca2fa346221",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2020_12_001/d10_1016_j_stem_2020_12_001.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2020_12_001/d10_1016_j_stem_2020_12_001.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2033ee58-170e-416b-805a-7677a81e2101",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 12\n",
    "\n",
    "files = os.listdir(\"raw/d10_1038_s41563_020_00853_9/GSE152014_RAW/\")\n",
    "\n",
    "#files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1038_s41563_020_00853_9/GSE152014_RAW/\"\n",
    "study = \"d10_1038_s41563_020_00853_9\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath + prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    #adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Howden_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d25\"\n",
    "    #adata.obs['study_id'] = \"6\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "\n",
    "adata.write(\"raw/d10_1038_s41563_020_00853_9/d10_1038_s41563_020_00853_9.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cec1300-327d-40a7-ba65-687cc1a56bc8",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1038_s41563_020_00853_9/d10_1038_s41563_020_00853_9.h5ad\",\n",
    "                          output_file=\"raw/d10_1038_s41563_020_00853_9/d10_1038_s41563_020_00853_9.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "0da20294-675e-46b1-b95d-b2f1405b2fc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct feature files\n",
    "import glob\n",
    "import gzip\n",
    "\n",
    "basepath = \"raw/d10_1038_s41467_019_13382_0/GSE136314_RAW/\"\n",
    "feature_files = glob.glob(os.path.join(basepath, \"*_features.tsv.gz\"))\n",
    "\n",
    "# Process each file\n",
    "for file in feature_files:\n",
    "    # Read features file\n",
    "    df = pd.read_csv(file, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Add new column \"Gene Expression\"\n",
    "    df['1'] = df.iloc[:, 0]\n",
    "    df[\"Gene Expression\"] = \"Gene Expression\"\n",
    "    \n",
    "    # Write back to gzipped file without adding extra newlines\n",
    "    with gzip.open(file, \"wt\", encoding=\"utf-8\") as f:\n",
    "        df.to_csv(f, sep=\"\\t\", header=False, index=False, lineterminator=\"\\n\")  # Ensures no extra newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6d6a1b0-7be4-42d3-b77c-427a18438bcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 13\n",
    "\n",
    "files = os.listdir(\"raw/d10_1038_s41467_019_13382_0/GSE136314_RAW/\")\n",
    "\n",
    "files = set([\"_\".join(x.split(\"_\")[:4]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1038_s41467_019_13382_0/GSE136314_RAW/\"\n",
    "study = \"d10_1038_s41467_019_13382_0\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Takasato_2016\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"HiSeq_X_Ten\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM4044536\": \"d29\", \"GSM4044537\": \"d15\", \"GSM4044538\": \"d7\", \"GSM4044539\": \"d0\", \"GSM4044540\": \"d29\", \n",
    "          \"GSM4044541\":\"d15\", \"GSM4044542\":\"d7\", \"GSM4044543\":\"d0\", \"GSM4044544\":\"d32\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "\n",
    "adata.write(\"raw/d10_1038_s41467_019_13382_0/d10_1038_s41467_019_13382_0.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8114f26-ad7b-44c2-924b-211e69aa8787",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1038_s41467_019_13382_0/d10_1038_s41467_019_13382_0.h5ad\",\n",
    "                          output_file=\"raw/d10_1038_s41467_019_13382_0/d10_1038_s41467_019_13382_0.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9931f9f-2bb6-484b-936c-c918d671abd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 14\n",
    "\n",
    "files = os.listdir(\"raw/d10_1016_j_devcel_2019_06_001/GSE124472_RAW/\")\n",
    "\n",
    "#files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1016_j_devcel_2019_06_001/GSE124472_RAW/\"\n",
    "study = \"d10_1016_j_devcel_2019_06_001\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath + prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Tran_2019\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v2\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NextSeq_500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM3534660\": \"d16\", \"GSM3534661\": \"d16\", \"GSM3534662\": \"d28\", \"GSM3534663\": \"d28\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_devcel_2019_06_001/d10_1016_j_devcel_2019_06_001.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ccdff425-6018-4725-8cca-c0d314259cc0",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_devcel_2019_06_001/d10_1016_j_devcel_2019_06_001.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_devcel_2019_06_001/d10_1016_j_devcel_2019_06_001.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0a837de7-25c8-4900-9d8b-6b2037a96bfe",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Correct feature files\n",
    "import glob\n",
    "import gzip\n",
    "\n",
    "basepath = \"raw/d10_15252_embr_201847483/GSE119561_RAW/\"\n",
    "feature_files = glob.glob(os.path.join(basepath, \"*_features.tsv.gz\"))\n",
    "\n",
    "# Process each file\n",
    "for file in feature_files:\n",
    "    # Read features file\n",
    "    df = pd.read_csv(file, sep=\"\\t\", header=None)\n",
    "\n",
    "    # Add new column \"Gene Expression\"\n",
    "    df[\"Gene Expression\"] = \"Gene Expression\"\n",
    "    \n",
    "    # Write back to gzipped file without adding extra newlines\n",
    "    with gzip.open(file, \"wt\", encoding=\"utf-8\") as f:\n",
    "        df.to_csv(f, sep=\"\\t\", header=False, index=False, lineterminator=\"\\n\")  # Ensures no extra newlines"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "6a5093b2-0fd6-42c4-94f0-e085697d12d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample GSM3377671_E6_ is processed\n",
      "Sample GSM3377672_SIX2_ is processed\n"
     ]
    }
   ],
   "source": [
    "# 15\n",
    "\n",
    "files = os.listdir(\"raw/d10_15252_embr_201847483/GSE119561_RAW/\")\n",
    "\n",
    "files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_15252_embr_201847483/GSE119561_RAW/\"\n",
    "study = \"d10_15252_embr_201847483\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Takasato_2016\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v2\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    #adata.obs['Age'] = \"\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM3377671\": \"d25\", \"GSM3377672\": \"d18\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "\n",
    "adata.write(\"raw/d10_15252_embr_201847483/d10_15252_embr_201847483.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba43bb2-c919-4970-8992-303b452a1147",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_15252_embr_201847483/d10_15252_embr_201847483.h5ad\",\n",
    "                          output_file=\"raw/d10_15252_embr_201847483/d10_15252_embr_201847483.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b3bd9343-1ee5-45b3-bdb5-2e99f3120934",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample GSE114802_org4_ is processed\n",
      "Sample GSE114802_org_ is processed\n"
     ]
    }
   ],
   "source": [
    "# 16\n",
    "\n",
    "files = os.listdir(\"raw/d10_1038_s41592_018_0253_2/\")\n",
    "\n",
    "files = set([\"_\".join(x.split(\"_\")[:2]) + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1038_s41592_018_0253_2/\"\n",
    "study = \"d10_1038_s41592_018_0253_2\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Takasato_2016\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v2\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1038_s41592_018_0253_2/d10_1038_s41592_018_0253_2.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "123f5d0b-6fed-4f74-86ec-299d029ce610",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1038_s41592_018_0253_2/d10_1038_s41592_018_0253_2.h5ad\",\n",
    "                          output_file=\"raw/d10_1038_s41592_018_0253_2/d10_1038_s41592_018_0253_2.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "8b32421e-f0f7-4fa0-902c-381502184de3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 17\n",
    "import scipy.sparse as sp\n",
    "\n",
    "# Path to your folder containing .txt files\n",
    "folder_path = \"raw/d10_1016_j_stem_2018_10_010/GSE118184/\"\n",
    "\n",
    "# Get a list of all .txt files in the folder\n",
    "txt_files = [f for f in os.listdir(folder_path) if f.endswith(\".txt\")]\n",
    "\n",
    "# List to store individual AnnData objects\n",
    "adata_list = []\n",
    "\n",
    "# Process each file\n",
    "for file in txt_files:\n",
    "    file_path = os.path.join(folder_path, file)\n",
    "    \n",
    "    # Extract filename without extension\n",
    "    sample_name = file\n",
    "\n",
    "    # Read the .txt file (assumes tab-separated format)\n",
    "    df = pd.read_csv(file_path, sep=\"\\t\", index_col=0)\n",
    "    df = df.T \n",
    "    \n",
    "    # Convert to sparse matrix for efficient storage\n",
    "    X = sp.csr_matrix(df.values)  \n",
    "\n",
    "    # Create AnnData object\n",
    "    adata = sc.AnnData(X=X)\n",
    "\n",
    "    # Assign cell names (now in rows)\n",
    "    adata.obs_names = df.index  # Cells\n",
    "\n",
    "    # Assign gene names (now in columns)\n",
    "    adata.var_names = df.columns  # Genes\n",
    "\n",
    "    # Add filename as metadata column\n",
    "    adata.obs[\"sample_id\"] = sample_name  \n",
    "\n",
    "    # Store in the list\n",
    "    adata_list.append(adata)\n",
    "\n",
    "adata = sc.concat(adata_list, join='outer')\n",
    "adata.obs['diff_protocol'] = adata.obs['sample_id'].str.split('_').str[1]\n",
    "adata.obs['source'] = adata.obs['sample_id'].str.split('_').str[2]\n",
    "adata.obs['condition'] = adata.obs['sample_id'].str.split('_').str[3]\n",
    "adata.obs['sc_protocol'] = \"Dropseq\"\n",
    "adata.obs[\"disease\"] = \"Control\"\n",
    "adata.obs['type'] = \"scRNAseq\"\n",
    "adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "adata.obs[\"study\"] = 'd10_1016_j_stem_2018_10_010'\n",
    "adata.obs['Age'] = adata.obs.index.str.split('_').str[0]\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSE118184_Takasato_ES_Batch1.2_dge.txt\":\"d26\", \"GSE118184_Morizane_iPS_Batch1.2_dge.txt\":\"d26\",\n",
    "          \"GSE118184_Takasato_iPS_Batch1.2_dge.txt\":\"d26\", \"GSE118184_Morizane_ES_Batch1_dge.txt\":\"d26\", \"GSE118184_Takasato_iPS_day34_dge.txt\":\"d34\",\n",
    "          \"GSE118184_Takasato_iPS_Batch3_dge.txt\":\"d26\", \"GSE118184_Takasato_iPS_BDNF.inhibitor_dge.txt\":\"d26\", \"GSE118184_Morizane_iPS_Batch3_dge.txt\":\"d26\"}\n",
    "\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "# Update the 'Age' column based on index values\n",
    "adata.obs.loc[adata.obs.index.str.startswith(\"Day\"), \"Age\"] = adata.obs.index[adata.obs.index.str.startswith(\"Day\")]\n",
    "adata.obs['Age'] = adata.obs['Age'].str.replace(r'_.*', '', regex=True)\n",
    "adata.obs['Age'] = adata.obs['Age'].str.replace(r'Day', 'd', regex=True)\n",
    "\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2018_10_010/d10_1016_j_stem_2018_10_010.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627c773e-4bed-4c3f-b1a8-46d1d3d4f27f",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2018_10_010/d10_1016_j_stem_2018_10_010.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2018_10_010/d10_1016_j_stem_2018_10_010.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d155a0f3-532f-4a27-bb28-058af4150a1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 18\n",
    "\n",
    "file_path = \"raw/d10_1016_j_stem_2018_04_022/GSE109718_RAW/\"\n",
    "study = \"d10_1016_j_stem_2018_04_022\"\n",
    "\n",
    "\n",
    "adatas = []\n",
    "# Read count matrix (assuming tab-separated values)\n",
    "for names in os.listdir(file_path):\n",
    "    df = pd.read_csv(file_path + names, sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    adata = sc.AnnData(df.T)\n",
    "    adata.obs[\"sample_id\"] = names\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id]\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC-ESC\"\n",
    "    adata.obs['diff_protocol'] = \"Freedman\"\n",
    "    adata.obs['sc_protocol'] = \"Drop-seq\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d21\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_stem_2018_04_022/d10_1016_j_stem_2018_04_022.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0a88203-292e-42d3-a2f0-e64848b28fe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_stem_2018_04_022/d10_1016_j_stem_2018_04_022.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_stem_2018_04_022/d10_1016_j_stem_2018_04_022.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ffeaa7f1-535d-448c-a494-eb3a944c6df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 19\n",
    "\n",
    "file_path = \"raw/d10_1172_jci_insight_122697/GSE115986_RAW/\"\n",
    "study = \"d10_1172_jci_insight_122697\"\n",
    "\n",
    "adatas = []\n",
    "# Read count matrix (assuming tab-separated values)\n",
    "for names in os.listdir(file_path):\n",
    "    df = pd.read_csv(file_path + names, sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    adata = sc.AnnData(df.T)\n",
    "    adata.obs[\"sample_id\"] = names\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id]\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"ESC\"\n",
    "    adata.obs['diff_protocol'] = \"Freedman\"\n",
    "    adata.obs['sc_protocol'] = \"Drop-seq\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1172_jci_insight_122697/d10_1172_jci_insight_122697.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "72e47163-8f5d-41c2-85b1-44cc98b9fba5",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1172_jci_insight_122697/d10_1172_jci_insight_122697.h5ad\",\n",
    "                          output_file=\"raw/d10_1172_jci_insight_122697/d10_1172_jci_insight_122697_harmonized.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "29f5008c-d03e-455f-9bf0-d9c868c25dce",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample GSM3273570_ is processed\n"
     ]
    }
   ],
   "source": [
    "# 20\n",
    "\n",
    "files = os.listdir(\"raw/d10_1242_dev_172361/GSE117211_RAW/\")\n",
    "\n",
    "files = set([x.split(\"_\")[0] + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1242_dev_172361/GSE117211_RAW/\"\n",
    "study = \"d10_1242_dev_172361\"\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Takasato_modified\"\n",
    "    adata.obs['sc_protocol'] = \"10x_v2\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata.write(\"raw/d10_1242_dev_172361/d10_1242_dev_172361.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aaeccfdf-bdf0-45c5-8e78-613b3df708c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1242_dev_172361/d10_1242_dev_172361.h5ad\",\n",
    "                          output_file=\"raw/d10_1242_dev_172361/d10_1242_dev_172361.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "a4d1bc59-ff1d-41ae-9cf6-6cc1c59b71ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 21\n",
    "\n",
    "file_path = \"raw/d10_1016_j_celrep_2020_108514/GSE131086_RAW/\"\n",
    "study = \"d10_1016_j_celrep_2020_108514\"\n",
    "\n",
    "adatas = []\n",
    "# Read count matrix (assuming tab-separated values)\n",
    "for names in os.listdir(file_path):\n",
    "    df = pd.read_csv(file_path + names, sep=\"\\t\", index_col=0, compression=\"gzip\")\n",
    "    adata = sc.AnnData(df.T)\n",
    "    adata.obs[\"sample_id\"] = names\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id]\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\"\n",
    "    adata.obs['diff_protocol'] = \"Takasato_Uchimura\"\n",
    "    adata.obs['sc_protocol'] = \"10X_5\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d26\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "adata.write(\"raw/d10_1016_j_celrep_2020_108514/d10_1016_j_celrep_2020_108514.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08b66f79-9533-404c-a163-553f82b7ad2b",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1016_j_celrep_2020_108514/d10_1016_j_celrep_2020_108514.h5ad\",\n",
    "                          output_file=\"raw/d10_1016_j_celrep_2020_108514/d10_1016_j_celrep_2020_108514.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "3ca106a2-914a-40f9-9067-c4489710e1fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sample GSM5032947_ is processed\n"
     ]
    }
   ],
   "source": [
    "# 22\n",
    "\n",
    "files = os.listdir(\"raw/d10_1186_s13073_022_01023_z/GSE165408_RAW/\")\n",
    "\n",
    "files = set([x.split(\"_\")[0] + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/d10_1186_s13073_022_01023_z/GSE165408_RAW/\"\n",
    "study = \"d10_1186_s13073_022_01023_z\"\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix # Add metadata to keep track of samples\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id] \n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\" #  iPSC\n",
    "    adata.obs['diff_protocol'] = \"Takasato_modified\"\n",
    "    adata.obs['sc_protocol'] = \"10x_v2\"\n",
    "    adata.obs['sequencing'] = \"Illumina_HiSeq_2500\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['Age'] = \"d25\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata.write(\"raw/d10_1186_s13073_022_01023_z/d10_1186_s13073_022_01023_z.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17195ed9-3b86-47ed-aa9e-da600b96cc0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1186_s13073_022_01023_z/d10_1186_s13073_022_01023_z.h5ad\",\n",
    "                          output_file=\"raw/d10_1186_s13073_022_01023_z/d10_1186_s13073_022_01023_z.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f850356-3afd-409b-ab45-a6cc148112ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 23 LMRH\n",
    "\n",
    "files = os.listdir(\"raw/dno_doi_kidney_organoid_LMRH/LMRH/\")\n",
    "\n",
    "files = set([x.split(\"_\")[0] + \"_\" for x in files])\n",
    "\n",
    "basepath = \"raw/dno_doi_kidney_organoid_LMRH/LMRH/\"\n",
    "study = \"dno_doi_kidney_organoid_LMRH\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_mtx(\n",
    "        basepath,\n",
    "        var_names=\"gene_symbols\",\n",
    "        prefix= prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id]\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\"\n",
    "    adata.obs['diff_protocol'] = \"Morizane_modified\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['type'] = \"scRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"2G9.R2\": \"d22\", \"2G9.R1\": \"d22\", \"12gneg\": \"d25\", \"12g.neg.j38\": \"d38\", \"05.4\": \"d25\",\n",
    "          \"WIDGET1\":\"d25\", \"WIDGET2\":\"d25\", \"WIDGET3\":\"d25\", \"WIDGET.NT\":\"d25\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "adata.write(\"raw/dno_doi_kidney_organoid_LMRH/dno_doi_kidney_organoid_LMRH.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "2b2237c2-fe4a-4ede-8965-00a31ae7e6b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/dno_doi_kidney_organoid_LMRH/dno_doi_kidney_organoid_LMRH.h5ad\",\n",
    "                          output_file=\"raw/dno_doi_kidney_organoid_LMRH/dno_doi_kidney_organoid_LMRH.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d17b5f73-4ed1-4255-8d82-51bb595865f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 24 \n",
    "\n",
    "files = os.listdir(\"raw/d10_1073_pnas_2219699120/raw/\")\n",
    "\n",
    "basepath = \"raw/d10_1073_pnas_2219699120/raw/\"\n",
    "study = \"d10_1073_pnas_2219699120\"\n",
    "\n",
    "adatas = []\n",
    "\n",
    "for prefix in files:\n",
    "    adata = sc.read_10x_h5(\n",
    "        basepath + prefix\n",
    "    )\n",
    "    adata.obs_names_make_unique()\n",
    "    adata.var_names_make_unique()\n",
    "    adata.obs[\"sample_id\"] = prefix\n",
    "    adata.obs[\"sample_id\"] = [x.split(\"_\")[0] for x in adata.obs.sample_id]\n",
    "    adata.obs[\"study\"] = study\n",
    "    adata.obs['source'] = \"hiPSC\"\n",
    "    adata.obs['diff_protocol'] = \"Takasato_modified\"\n",
    "    adata.obs['sc_protocol'] = \"10x_3_v3.1\"\n",
    "    adata.obs['sequencing'] = \"Illumina_NovaSeq_6000\"\n",
    "    adata.obs['genome_build'] = \"hg38\"\n",
    "    adata.obs['type'] = \"snRNAseq\"\n",
    "    adata.obs[\"disease\"] = \"Control\"\n",
    "    adatas.append(adata)\n",
    "    print(f\"Sample {prefix} is processed\")\n",
    "\n",
    "\n",
    "adata = sc.concat(adatas, join='outer')\n",
    "\n",
    "# Define the mapping dictionary\n",
    "mapping = {\"GSM6573624\": \"d26\", \"GSM6573626\": \"d26\", \"GSM6573628\": \"d26\", \"GSM6573630\": \"d19\", \"GSM6573634\": \"d16\", \"GSM6573636\":\"d12\",\n",
    "          \"GSM6573638\":\"d12\", \"GSM6573640\":\"d7\", \"GSM6573642\":\"d7\"}\n",
    "\n",
    "# Create the new column based on `sample_id`\n",
    "adata.obs[\"Age\"] = adata.obs[\"sample_id\"].map(mapping)\n",
    "\n",
    "\n",
    "adata.write(\"raw/d10_1073_pnas_2219699120/d10_1073_pnas_2219699120.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e0456a57-879f-4e1b-af4c-b4e6afbb83dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "memory_efficient_harmonize(input_file=\"raw/d10_1073_pnas_2219699120/d10_1073_pnas_2219699120.h5ad\",\n",
    "                          output_file=\"raw/d10_1073_pnas_2219699120/d10_1073_pnas_2219699120.h5ad\",\n",
    "                           harmonize_matrix= harmonize_matrix)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
